src/storage/mod.rs

```diff
@@ -1595,83 +1641,100 @@ impl<E: Engine> Storage<E> {
+                                &snapshot,
+                                &cf,
+                                &start_key,
+                                end_key,
+                                each_limit,
+                                &mut statistics,
+                                key_only,
+                            )?
+                        } else {
+                            Self::raw_scan(
+                                &snapshot,
+                                &cf,
+                                &start_key,
+                                end_key,
+                                each_limit,
+                                &mut statistics,
+                                key_only,
+                            )?
+                        };
+                        result.extend(pairs.into_iter());
+                    }
 
-                        thread_ctx.collect_read_flow(region_id, &statistics);
-                        thread_ctx.collect_key_reads(CMD, statistics.write.flow_stats.read_keys as u64);
-                        thread_ctx.collect_scan_count(CMD, &statistics);
+                    thread_ctx.collect_read_flow(region_id, &statistics);
+                    thread_ctx.collect_key_reads(CMD, statistics.write.flow_stats.read_keys as u64);
+                    thread_ctx.collect_scan_count(CMD, &statistics);
 
-                        Ok(result)
-                    })
-                    .and_then(move |res| {
-                        timer.observe_duration();
-                        let r = tx.send(res);
-                        if r.is_err() {
-                            warn!("future_callback: Failed to send result to the future rx, discarded.");
-                        }
-                        Ok(())
-                    })
-            }) {
-                fut.forget();
+                    Ok(result)
+                })
+                .and_then(move |res| {
+                    timer.observe_duration();
+                    let r = tx.send(res);
+                    if r.is_err() {
+                        warn!(
+                            "future_callback: Failed to send result to the future rx, discarded."
+                        );
+                    }
+                    Ok(())
+                })
+        }) {
+            Ok(res) => res.forget(),
+            Err(_e) => {
+                let _ = tx2.send(Error::SchedTooBusy);
             }
         };
 
-        let res = engine.async_snapshot(&ctx, cb);
-        future::result(res)
-            .map(|_| future.map_err(|e| EngineError::Other(box_err!(e))))
-            .flatten()
-            .map_err(|_| Error::SchedTooBusy)
+        let _res = engine.async_snapshot(&ctx, cb);
+
+        future
+            .select2(future2)
+            .then(|res| -> Box<dyn Future<Item = _, Error = _> + Send> {
+                match res {
+                    Ok(future::Either::A((x, _))) => Box::new(future::result(Ok(x))),
```

Comment:
Looks quite complicated. Can we just:

```rust
future::result(future2)
    .and_then(|_| future::result(res).map_err(|e| EngineError::Other(box_err!(e))))
    .flatten()
    ....
```

Reply:
The `select2` way contains all possible situations of two channels (`tx`->`future`) and (`tx2`->`future2`):
- path-1,  `tx` sends successful firstly, just get the wanted result `Option<Vec<u8>>` ;
- path-2,  `tx2` sends successful firstly, here `readpool` occurs `Full` error, then just return `Error::SchedTooBusy`;
- path-3, `tx` dropped firstly, then poll `future2`, here two situations: 1) `tx2` sends successful, i.e. occurs Full error, then return like path-2; 2) `tx2` also dropped secondly, then return `Canceled` error;
- path-4, `tx2` dropped firstly, then poll `future`, here two situations: 1) `tx1` sends successful, then return like path-1; 2) `tx` also dropped secondly, then return `Canceled` error.
From test, normal flow is path-4: `tx2` dropped firstly, then `tx` sends successful, get the wanted result.  If readpool is full, go path-2 or path-3.